File: frontend/lib/sections.ts
import type { ProfileAsset, ProfileSection, SectionType } from "./types";

export const MAX_SECTIONS = 12;

export const SECTION_CATALOG: Array<{ id: SectionType; label: string; hint: string }> = [
  { id: "about", label: "About me", hint: "Markdown bio under your card." },
  { id: "project", label: "Project", hint: "Cover, tags, and a link." },
  { id: "skills", label: "Skills", hint: "A row of skill chips." },
  { id: "text", label: "Custom text", hint: "A free Markdown block." },
  { id: "lyrics", label: "Synced lyrics", hint: "LRC lines follow the music player." },
];

const SECTION_TYPES = new Set(SECTION_CATALOG.map((item) => item.id));

export function sectionLabel(type: SectionType) {
  return SECTION_CATALOG.find((item) => item.id === type)?.label || type;
}

export function createSectionId(type: SectionType) {
  const unique = typeof crypto !== "undefined" && "randomUUID" in crypto
    ? crypto.randomUUID()
    : `${Date.now()}-${Math.random().toString(36).slice(2, 10)}`;
  return `${type}-${unique}`.slice(0, 40);
}

export function defaultSection(type: SectionType): ProfileSection {
  const titles: Record<SectionType, string> = {
    about: "About",
    project: "Project",
    skills: "Skills",
    text: "",
    lyrics: "Lyrics",
  };
  return { id: createSectionId(type), type, enabled: true, title: titles[type], body: "", href: "", tags: [], cover: { url: null } };
}

export function normalizeSections(raw: ProfileSection[] | undefined): ProfileSection[] {
  const seen = new Set<string>();
  return (raw || []).slice(0, MAX_SECTIONS).flatMap((item, index) => {
    if (!item || !SECTION_TYPES.has(item.type)) return [];
    let id = String(item.id || `${item.type}-${index + 1}`).slice(0, 40);
    if (seen.has(id)) id = `${id}-${index + 1}`.slice(0, 40);
    seen.add(id);
    const tags = Array.isArray(item.tags) ? item.tags.map((tag) => String(tag || "").trim()).filter(Boolean).slice(0, 16) : [];
    const cover = item.cover && typeof item.cover === "object" ? item.cover : { url: null };
    return [{
      id,
      type: item.type,
      enabled: Boolean(item.enabled),
      title: String(item.title || "").slice(0, 80),
      body: String(item.body || "").slice(0, 8000),
      href: String(item.href || "").slice(0, 500),
      tags,
      cover: { url: cover.url || null, name: cover.name || "", type: cover.type || "" },
    }];
  });
}

export function parseLyrics(body: string) {
  return (body || "").split(/\r?\n/).flatMap((raw) => {
    const text = raw.trim();
    if (!text) return [];
    const match = /^\[(\d{1,2}):(\d{2})(?:\.(\d{1,3}))?\]\s*(.*)$/.exec(text);
    if (match) {
      const stamp = Number(match[1]) * 60 + Number(match[2]) + Number((match[3] || "0").padEnd(3, "0").slice(0, 3)) / 1000;
      return [{ t: stamp, text: (match[4] || "").trim().slice(0, 200) }];
    }
    return [{ t: null as number | null, text: text.slice(0, 200) }];
  }).slice(0, 200);
}

export function tagsFromInput(value: string) {
  return value.split(",").map((item) => item.trim()).filter(Boolean).slice(0, 16);
}

export function tagsToInput(tags: string[] | undefined) {
  return (tags || []).join(", ");
}

export function sectionCoverSrc(username: string, section: ProfileSection, preview: boolean) {
  if (preview) return section.cover?.url || "";
  return section.cover?.url ? `/api/v1/profile/${encodeURIComponent(username)}/sections/${encodeURIComponent(section.id)}/cover` : "";
}

export function emptyCover(): ProfileAsset {
  return { url: null };
}

File: backend/app/core/sections.py
from typing import Any


def section_has_content(item: dict[str, Any]) -> bool:
    kind = str(item.get("type") or "")
    if kind in {"about", "integration"} and any((item.get(side) or {}).get("enabled") for side in ("leftCard", "rightCard")):
        return True
    if kind == "lyrics":
        return True
    if kind == "skills":
        return bool(item.get("tags"))
    if kind == "project":
        cover = item.get("cover") if isinstance(item.get("cover"), dict) else {}
        return bool(item.get("title") or item.get("body") or item.get("href") or cover.get("url"))
    return bool(str(item.get("body") or "").strip() or str(item.get("title") or "").strip() or item.get("subtitle") or item.get("tags"))


def project_configured(item: dict[str, Any]) -> bool:
    return bool(str(item.get("body") or "").strip() or str(item.get("href") or "").strip() or (item.get("cover") or {}).get("url") or item.get("tags"))


def visible_sections(items: list) -> list[dict]:
    enabled = [item for item in items if isinstance(item, dict) and item.get("enabled")]
    about = next((item for item in enabled if item.get("type") == "about"), None)
    skills = [tag for item in enabled if item.get("type") == "skills" for tag in item.get("tags") or []]
    result = []
    index = 0
    while index < len(enabled):
        item = enabled[index]
        index += 1
        kind = item.get("type")
        if kind == "skills" and (about is not None or not item.get("tags")):
            continue
        if item is about:
            item = {**item, "tags": list(dict.fromkeys([*(item.get("tags") or []), *skills]))}
        if kind == "project":
            group = [item]
            while index < len(enabled) and enabled[index].get("type") == "project":
                group.append(enabled[index])
                index += 1
            result.extend([entry for entry in group if project_configured(entry)] or group[:1])
        elif kind == "integration" or section_has_content(item):
            result.append(item)
    return result
